File: frontend/src/app/core/services/review.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { Review, ReviewRequest } from '../models/review.model';
import { ReorderItem } from '../models/reorder-item.model';

@Injectable({ providedIn: 'root' })
export class ReviewService {
  private readonly http = inject(HttpClient);

  list(): Observable<Review[]> {
    return this.http.get<Review[]>('/api/reviews');
  }

  create(request: ReviewRequest): Observable<Review> {
    return this.http.post<Review>('/api/admin/reviews', request);
  }

  update(id: number, request: ReviewRequest): Observable<Review> {
    return this.http.put<Review>(`/api/admin/reviews/${id}`, request);
  }

  reorder(items: ReorderItem[]): Observable<void> {
    return this.http.put<void>('/api/admin/reviews/reorder', items);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`/api/admin/reviews/${id}`);
  }
}

File: backend/src/main/kotlin/com/svadbarium/venue/review/ReviewService.kt
package com.svadbarium.venue.review

import com.svadbarium.venue.common.dto.ReorderItem
import com.svadbarium.venue.common.exception.NotFoundException
import com.svadbarium.venue.review.dto.ReviewDto
import com.svadbarium.venue.review.dto.ReviewRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ReviewService(
    private val repository: ReviewRepository,
) {
    fun list(): List<ReviewDto> = repository.findAllByOrderBySortOrderAsc().map { it.toDto() }

    @Transactional
    fun create(request: ReviewRequest): ReviewDto {
        val nextSortOrder = (repository.findAllByOrderBySortOrderAsc().maxOfOrNull { it.sortOrder } ?: -1) + 1
        val review = Review(
            reviewerName = request.reviewerName,
            reviewText = request.reviewText,
            reviewDate = request.reviewDate,
            googleReviewUrl = request.googleReviewUrl,
            sortOrder = nextSortOrder,
        )
        return repository.save(review).toDto()
    }

    @Transactional
    fun update(id: Long, request: ReviewRequest): ReviewDto {
        val review = repository.findById(id).orElseThrow { NotFoundException("Review $id not found") }
        review.reviewerName = request.reviewerName
        review.reviewText = request.reviewText
        review.reviewDate = request.reviewDate
        review.googleReviewUrl = request.googleReviewUrl
        return repository.save(review).toDto()
    }

    @Transactional
    fun reorder(items: List<ReorderItem>) {
        val reviews = repository.findAllById(items.map { it.id }).associateBy { it.id }
        items.forEach { item ->
            reviews[item.id]?.sortOrder = item.sortOrder
        }
        repository.saveAll(reviews.values)
    }

    @Transactional
    fun delete(id: Long) {
        if (!repository.existsById(id)) {
            throw NotFoundException("Review $id not found")
        }
        repository.deleteById(id)
    }

    private fun Review.toDto() = ReviewDto(
        id = id,
        reviewerName = reviewerName,
        reviewText = reviewText,
        reviewDate = reviewDate,
        googleReviewUrl = googleReviewUrl,
        sortOrder = sortOrder,
    )
}
